kernel/fatfs.c

```diff
@@ -133,10 +134,45 @@ STATIC void init_direntry(struct dirent *dentry, unsigned attrib,
 int dos_open(char *path, unsigned flags, unsigned attrib, int fd)
 {
   REG f_node_ptr fnp = sft_to_fnode(fd);
-  int status = find_fname(path, D_ALL | attrib, fnp);
+  int status;
+
+  /* Special handling for volume name (they can coexist with files/directories)
+   */
+  if (((attrib & (D_DIR | D_VOLID | D_SYSTEM | D_HIDDEN | D_RDONLY)) == D_VOLID) && (flags & O_TRUNC)) {
```

Comment:
I read that it's required to set volume id as you can't open an existing label as a file, but I don't have access to my DOS book for the next few days to check. I did some tests in an earlier ticket which showed https://github.com/dosemu2/dosemu2/issues/1771 there was very little you could do other than create the label (no others can exist on the volume), write, seek on it, and close. Afterwards all you could do was find and delete it.

Reply:
What about O_CREAT?

Reply:
Yes that would seem more likely than O_TRUNC, especially as the caller has to use the int21/16 FCB create/truncate for labels. https://fd.lod.bz/rbil/interrup/dos_kernel/2116.html. I'll see about printing out what bits are set when label is created.

Reply:
Edited that last comment.

Reply:
So int21/16 call generates flags of
~~~
fdpp: System call (21h): 1600
fdpp: HELLO flags = 0x8e02 

# FDPP has flags defined as 
>>> #define O_NOINHERIT     0x0080                                                  
>>> #define O_OPEN          0x0100 /* not     */                                    
>>> #define O_TRUNC         0x0200 /*    both */                                    
>>> #define O_CREAT         0x0400                                                  
>>> #define O_LEGACY        0x0800                                                  
>>> #define O_LARGEFILE     0x1000                                                  
>>> #define O_NOCRIT        0x2000                                                  
>>> #define O_SYNC          0x4000                                                  
>>> #define O_FCB           0x8000                                                  
~~~
So it looks like O_LEGACY | O_CREAT | O_TRUNC are set along with O_FCB       

I noticed I hadn't done any tests that use int21/0f to open, so as yet I have no value for the flags in that case.

However I just spotted a problem with label create, it fails if there is a previously deleted label on the disk (first char is 0xe5) and has the volid set. If I use a disk editor (https://github.com/Baron-von-Riedesel/WDe/releases/tag/v1.0) to remove that bit it is able to successfully create a new one.

Reply:
Fails because find_in_dir() finds that?

Reply:
```
  swap_deleted(fnp->f_dir.dir_name);

  /* and for efficiency, stop when we hit the first       */
  /* unused entry.                                        */
  /* either returns 1 or 0                                */
  return (fnp->f_dir.dir_name[0] != '\0');
```
Looks like this code is broken, can you
make any sense of it?

Reply:
> Fails because find_in_dir() finds that?

Not sure yet, I just wrote a test. My previous test didn't uncover the problem as the test framework always starts with pristine conditions to ensure repeatability Only when I ran the test by hand a second time did I see the problem.

> .Looks like this code is broken, can you
make any sense of it?

Wouldn't that fragment mean that deleted directory entries aren't reused?

Reply:
> Wouldn't that fragment mean that deleted directory entries aren't reused?

Likely not.
They test for `\0` instead of '\5'
or '\e5'. e5 is EXT_DELETED, no
idea what does it actually mean.
And if you change that code to test
for 5 or e5, then it will return 0 and
abort the search on the first unused
entry, which is definitely undesirable.
No idea what's going on here.

Reply:
I have to go out now for a while, but the new test is there in https://github.com/andrewbird/dosemu2/tree/ci-01 and can be used as
~~~
test/test_dos.py PPDOSGITTestCase.test_fat_label_delete_recreate
~~~

Reply:
Added a possible fix, could you please check?

Reply:
So testing your latest fdpp plus my patches with that new test I get this:
~~~
test/test_dos.py PPDOSGITTestCase.test_fat_label_delete_recreate
~~~

~~~
C:\>testit.bat
testit.bat

C:\>d:
D:\>c:\labdelr
FAIL: On recreation

D:\>DIR
 Volume in drive d is ÕNITIAL
 Directory of d:\

File not found
~~~

Which looks to be a problem with delete, so I tested with the wildcard delete test
~~~
test/test_dos.py PPDOSGITTestCase.test_fat_label_delete_wildcard
~~~

~~~
C:\>testit.bat
testit.bat

C:\>d:
D:\>c:\labdelw
PASS: Operation success

D:\>DIR
 Volume in drive d is ÕNITIAL
 Directory of d:\

File not found
~~~

Which looks like the wildcard delete thinks it was successful, but for some reason DIR is showing a corrupted label when it should not be present.

Reply:
re-adding the `swap_deleted()` and making sure the loop was incremented fixed both tests for me
```diff
diff --git a/kernel/fatdir.c b/kernel/fatdir.c
index 975a55e..233a998 100644
--- a/kernel/fatdir.c
+++ b/kernel/fatdir.c
@@ -215,7 +215,7 @@ COUNT dir_read(REG f_node_ptr fnp)
   fnp->f_diridx = entry % (secsize / DIRENT_SIZE);
   getdirent(&bp->b_buffer[fnp->f_diridx * DIRENT_SIZE], &fnp->f_dir);
 
-//  swap_deleted(fnp->f_dir.dir_name);
+  swap_deleted(fnp->f_dir.dir_name);
 
   /* and for efficiency, stop when we hit the first       */
   /* unused entry.                                        */
diff --git a/kernel/fatfs.c b/kernel/fatfs.c
index 17b46fe..feb41d6 100644
--- a/kernel/fatfs.c
+++ b/kernel/fatfs.c
@@ -324,10 +324,10 @@ BOOL fcbmatch(const char *fcbname1, const char *fcbname2)
 
 STATIC int find_in_dir(int attr_req, int attr_allow, f_node_ptr fnp)
 {
-  while (dir_read(fnp) == 1)
+  for (/* */; dir_read(fnp) == 1; fnp->f_dmp->dm_entry++)
   {
     if (fnp->f_dir.dir_name[0] == DELETED)
-        continue;
+      continue;
     if ((fnp->f_dmp->dm_name_pat[0] == '\0'
         || fcbmatch(fnp->f_dir.dir_name, fnp->f_dmp->dm_name_pat))
         && ((fnp->f_dir.dir_attrib & attr_req) == attr_req)
@@ -335,7 +335,6 @@ STATIC int find_in_dir(int attr_req, int attr_allow, f_node_ptr fnp)
     {
       return SUCCESS;
     }
-    fnp->f_dmp->dm_entry++;
   }
   return DE_FILENOTFND;
 }
```

Reply:
Maybe @bartoldeman knows what
does this swap_deleted() do?

Reply:
Thanks, does something like this work
now? I only used EXT_DELETED, which
I suppose is needed after swap_deleted()
but you use DELETED, so unsure...

Reply:
Found other code uses DELETED too, not
EXT_DELETED. Where exactly have you
seen e5 then?

Reply:
swap_deleted swaps the external delete marker 0xe5 for the internal 0x05.
The reason for this is that if the on-disk byte is 0x05 it needs to be translated to 0xe5, crucially used in some Japanese character sets (also UTF-8 of course, but that came later).
See also https://github.com/torvalds/linux/blob/e01d50cbd6eece456843717a566a34e8b926cf0c/fs/fat/namei_msdos.c#L56

Reply:
Thanks, your latest tweak runs the label tests fine. I saw the 0xe5 in the disk editor
![Screenshot_2022-11-15_16-52-31](https://user-images.githubusercontent.com/220835/201979262-68a157a1-ba20-4be6-87ff-03d42e968eb0.png)

Reply:
> The reason for this is that if the on-disk byte is 0x05

In what case the on-disk byte is 05?
Andrew indicates on-disk byte after
deletion is e5.
Also I don't understand what swap_deleted()
does in `dir_write_update()`. It seems to be
called twice in a rather linear code block.

Actually I don't understand anything at all.
From your answer I assume there can be 2
different on-disk deletion markers, perhaps
05 is used by some older versions of DOS.
What it helps for, to swap one to another?
This means you still can get out of `dir_read()`
with either one, right?

Reply:
> Thanks, your latest tweak runs the label tests fine. I saw the 0xe5 in the disk editor

Thanks, so since I see the same code
in other places, I assume my current change
is correct, even if I fail to understand it no
matter how hard I try. :)
So please use it for your purposes.

Reply:
> In what case the on-disk byte is 05? Andrew indicates on-disk byte after deletion is e5. Also I don't understand what swap_deleted() does in `dir_write_update()`. It seems to be called twice in a rather linear code block.
> 
> Actually I don't understand anything at all. From your answer I assume there can be 2 different on-disk deletion markers, perhaps 05 is used by some older versions of DOS. What it helps for, to swap one to another? This means you still can get out of `dir_read()` with either one, right?

"00" means the directory entry was never used and there are no more used entries after it as well.

"E5" means the directory entry is deleted. Other directory entries may follow, however.

"05" (on disk) means that the entry is **not** a deleted entry but rather a normal entry whose first name byte is meant to be "E5" (in pathname), which cannot be stored literally because it would collide with the "E5" (on disk) meaning deleted. DOS wants to translate (Code Page dependent) "E5" (in pathname) codepoints to on-disk "05" if they occur in the first byte of the filename.

Reply:
OK I am starting to understand, thanks to
both of you. So calling swap_deleted() "in
a linear block" means "write 05 to disk and
restore name back to e5; hope 05 never
comes in a name initially or we will write a
deleted entry e5 then".
I would guess when e5 is read from disk,
we don't have to swap it to 05. We can as
well change it into 01? I'd really like to make
this code more readable.